proposals/time-stamp-counter-support.md

```diff
@@ -0,0 +1,274 @@
+    (typically the platform epoch). Used to aid in converting TscTimestamp 
+    instances to timestamps
+
+    start_tsc_timestamp: A tsc timestamp taken right after the timestamp 
+    syscall call returned. 
+
+    tsc_is_reliable: Whether the TSC is usable as a time source. If false, a monotonic but fast to query OS clock is used.
+    """
+
+    var freq: TscFrequency
+    var start_os_timestamp: UInt64
+    var start_tsc_timestamp: TscTimestamp
+    var tsc_is_reliable: Bool
+
+    fn __init__(inout self):
+        pass
+
+    fn get_frequency_hz(borrowed self) -> UInt64:
+        """
+        Get a TscFrequency object which can be used to get the time delta between
+        two TscTimestamp objects.
+        """
+        return self.freq.get_frequency_hz()
+
+    fn get_tsc_timestamp(borrowed self) -> TscTimestamp:
+        """
+        Read from the TSC (or fallback clock). 
+        """
+        pass
+
+    fn now(borrowed self) -> Time:
+        """
+        To be added once the standard library decides how to handle time.
+        """
+        pass
+```
+
+## Examples and Usecases
+
+### Benchmarking a hot loop
+
+This example shows a user gathering multiple datapoints per run, which is not
+possible with the current benchmark library. Using a system call here would
+cause `do_operation_2` to be started later than it otherwise would have,
+causing `end` to deliver unreliable numbers.
+
+```mojo
+fn hot_fn():
+    var tsc = Tsc()
+
+    var middle_durations: List[Int64] = List[Int64]()
+    var durations: List[Int64] = List[Int64]()
+
+    for i in range(10):
+        var start = tsc.get_tsc_timestamp()
+        var phase1 = do_operation_1()
+        var middle = tsc.get_tsc_timestamp()
+        do_operation_2(phase1)
+        var end = tsc.get_tsc_timestamp()
+
+        middle_durations.append(tsc.get_time_delta(start, middle))
+        durations.append(tsc.get_time_delta(start, end))
+```
+
+### Timestamps when logging
+
+When outputting a lot of logs, timestamps are frequently a large fraction of
+the cost to the degree that some logging libraries will actually treat whole
+batches of messages as occuring at the same instant to save on system calls.
+Using a TSC can help mitigate that by removing extra system calls entirely.
+
+```mojo
+struct Logger:
+    var tsc: Tsc
+    var prefix: String
+
+    fn log(borrowed self, borrowed message: String):
+        print(self.prefix, self.tsc.now(), " ", message)
+```
+
+### High precisions latency measurements
+
+In networked applications it is often desirable to obtain as precise as
+possible of a latency measurement. This is used for setting service level
+agreements (SLAs) as well as for reasoning about the correctness of some
+more exotic systems such as Google's Spanner database. The latency test
+harnesses for these systems often have hundreds or thousands of messages
+in flight at a time and want to make getting a timestamp as cheap as possible.
+
+## Why the Standard Library?
```

Comment:
+1 for putting it in Mojo standard library.  There's some prior art with Folly for example [here](https://github.com/facebook/folly/blob/9c9430777ac2041a3d9faa9d87862f1e35e42d19/folly/chrono/Hardware.h)